### DISCUSSION TITLE
Embed Rhai-rs in Glicol for sample-level music/DSP live coding in web browsers

### DISCUSSION THREAD

**Discussion by chaosprint** [rhaiscript/rhai#527] [Show and tell]
first, I would like to thank all the contributors of rhai especially @schungx

### TL;DR
I have been exploring the possibility to embed a script language into my project glicol (graph-oriented live coding language) is a computer music language and an audio DSP lib written in Rust:
```
o: sin 440 >> mul 0.2 // hello world
```
with rhai script, you can now write code like this to make music in web browsers:
```
o: script `
    output.pad(128, 0.0);
    for i in 0..128 {
        output[i] = sin(2*PI()*phase) ;
        phase += 440.0 / sr;
    };
    while phase > 1.0 { phase -= 1.0 };
    output
` >> script `
    output = input.map(|i|i*0.2);
    output
`
```
- try it on: [https://glicol.org](https://glicol.org)
- source code: [https://github.com/chaosprint/glicol](https://github.com/chaosprint/glicol)

---

### a little bit more on the process

one limitation for Glicol before, and perhaps in many other computer music languages/live coding languages, is that there is no mechanism to support sample-level dsp live coding. in Max/MSP, there is an `object` called `~gen` which is probably the most well-known for this purpose.

I have been looking for options for this embedding language and finally landed on `rhai`.

the syntax of `rhai` is really intuitive and I think it has merged the syntax sugar of `js` and `rust`.

embedding Rhai in Glicol is a very straightforward experience thanks to its well-designed api and docs ([rhai.rs](rhai.rs))

now, users can write:
```
a: script `
    output.pad(128, 0.0);
    for i in 0..128 {
        output[i] = sin(2*PI()*phase) ;
        phase += 440.0 / sr;
    };
    while phase > 1.0 { phase -= 1.0 };
    output
` >> script `
    output = input.map(|i|i*0.2);
    output
`
```
see the interactive example on [https://glicol.org/guide#rhai2](https://glicol.org/guide#rhai2)

from my current testing, for this real-time audio scripting, creating new variables, especially arrays should be avoided.

for now the `script` node provides `outout`, `input` (if there is one), `sr` and `phase` to the `Scope` of rhai, as well as variables from `a` to `z`. `x0`, `x1`, `x2`, `y0`, `y1`, `y2`.

I am still exploring the possibility to share a `ringbuf` in rhai `Scope`; some functions such as random can also be useful.

if you are interested, please give me some feedback here or on the github repo of glicol:
[https://github.com/chaosprint/glicol](https://github.com/chaosprint/glicol)

cheers!

**Comment by schungx** [comment]
> some functions such as random can also be useful.

There is a [`rhai-rand`](https://crates.io/crates/rhai-rand) crate.  :-)

**Comment by schungx** [comment]
> creating new variables, especially arrays should be avoided.

Yes, Rhai arrays hold `Dynamic` values.  There are no "typed arrays" in Rhai (other than BLOBs which is a byte array).

If you need a typed array, you'd need to register its API.

**Comment by schungx** [comment]
Closing this for now. Feel free to reopen if necessary.

**Comment by schungx** [comment]
Closing this for now. Feel free to reopen if necessary.
